ReSharper.FSharp/src/FSharp.Psi.Features/src/Intentions/FunctionAnnotationAction.fs

```diff
@@ -0,0 +1,133 @@
+﻿namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Intentions
+
+open FSharp.Compiler.SourceCodeServices
+open JetBrains.ReSharper.Feature.Services.ContextActions
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Util
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Impl
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Parsing
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Parsing
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Parsing
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Tree
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Util
+open JetBrains.ReSharper.Plugins.FSharp.Util
+open JetBrains.ReSharper.Psi
+open JetBrains.ReSharper.Psi.DataContext
+open JetBrains.ReSharper.Psi.ExtensionsAPI
+open JetBrains.ReSharper.Psi.ExtensionsAPI.Tree
+open JetBrains.ReSharper.Psi.Tree
+open JetBrains.ReSharper.Resources.Shell
+
+type private ParameterToModify = {
+    ParameterNameNode : ILocalReferencePat
+    ParameterNodeInSignature : ISynPat
+    FSharpParameters : FSharpParameter list
+}
+
+type private ParameterNodeWithType = {
+    Node : ITreeNode
+    FSharpParameters : FSharpParameter list
+}
+
+[<ContextAction(Name = "AnnotateFunction", Group = "F#", Description = "Annotate function with parameter types and return type")>]
+type FunctionAnnotationAction(dataProvider: FSharpContextActionDataProvider) =
+    inherit ContextActionBase()
+
+    let rec getTypeDisplayName (fsType : FSharpType) : string =
+        let typeDef = if fsType.HasTypeDefinition then fsType.TypeDefinition.DisplayName else ""
+        let genericParam = if fsType.IsGenericParameter then "'" + fsType.GenericParameter.DisplayName else ""
+        let genericArgs = if fsType.HasTypeDefinition then fsType.GenericArguments |> Seq.map getTypeDisplayName |> Seq.toList else List.empty
+        let tupleArgs = if fsType.IsTupleType then fsType.GenericArguments |> Seq.map getTypeDisplayName |> String.concat "*" else ""
+        
+        [genericParam; typeDef; tupleArgs]
+        |> List.append genericArgs
+        |> List.filter (fun x -> String.length x <> 0)
+        |> String.concat " "
+    
+    override x.Text = "Annotate function with parameter types and return type"
+    
+    override x.IsAvailable _ =
+        let letBindings = dataProvider.GetSelectedElement<ILetBindings>()
```

Comment:
If making this action work on a binding, it should be an `IBinding` element here.